✨ feat(git-commit): Add AI-powered git commit message generation

hassoncs opened the pull request:
I've been using a different extension to generate these messages, so might as well add it to kilocode!  

![2025-06-05 12 43 24](https://github.com/user-attachments/assets/e8d4508a-2592-48f9-8977-a00a816cbf6c)

*Recent commit edit* ai updated the button to be our KILO logo:
![image](https://github.com/user-attachments/assets/4a4adb4e-cb9b-4e3b-8d28-ec423312f020)


**Ticket / Discussion**
https://github.com/Kilo-Org/kilocode/discussions/620
https://github.com/Kilo-Org/kilocode/issues/623


**TODO eventually**
* Support customizing the prompt / provider (very similar to Autocomplete!)
* Support kilcode .rules files when generating commits

**Details**
  - Introduce `GitCommitMessageProvider` to generate commit messages using AI
  - Register new command `kilo-code.generateCommitMessage` in package.json
  - Add language support for new command title in localization files

✅ test(git-commit): add unit tests for GitCommitMessageProvider
  - Mock vscode and dependencies for testing GitCommitMessageProvider
  - Test activation, message generation, and error handling scenarios
  - Validate internal methods like `getChangeStatusText` and `cleanCommitMessage`

Commit:
✨ feat(git-commit): add AI-powered git commit message generation
  - Introduce `GitCommitMessageProvider` to generate commit messages using AI
  - Register new command `kilo-code.generateCommitMessage` in package.json
  - Add language support for new command title in localization files

✅ test(git-commit): add unit tests for GitCommitMessageProvider
  - Mock vscode and dependencies for testing GitCommitMessageProvider
  - Test activation, message generation, and error handling scenarios
  - Validate internal methods like `getChangeStatusText` and `cleanCommitMessage`

Commit:
✨ feat(commit-message): integrate custom rule loading for prompts

- add support for loading custom rules from workspace for commit message prompts
- modify buildCommitMessagePrompt to include additional rules if available

✅ test(commit-message): validate rule inclusion in commit message prompts

- mock loadRuleFiles to simulate rule loading in tests
- add test case to ensure rules are appended to prompts when available

Commit:
♻️ refactor(commit-message): rename GitUtils to GitExtensionService

- rename GitUtils to GitExtensionService for better clarity and alignment with its functionality
- update imports and references accordingly across the codebase

Commit:
♻️ refactor(tests): fix the tests

- rename MockGitUtils to MockGitExtensionService
- update all references to use the new mock name
- ensure consistency across test cases

Commit:
✨ feat(icons): add new kilo icons for theme support

- introduce kilo-dark.svg and kilo-light.svg to assets/icons
- support both dark and light themes for better visual integration

🔧 chore(package): update package.json with theme icons

- modify command icon to use new theme-specific SVGs in package.json
- enhance UI customization by linking light and dark icons

Commit:
💄 style(icons): use the kilo block icon for the button

In both light/dark modes

HadesArchitect commented:
Is that duplication intentional / needed? 
<img width="387" alt="image" src="https://github.com/user-attachments/assets/adab8efe-9b0d-4e1f-b014-9a1336cbcd37" />

HadesArchitect commented:
Suggestion: regenerate commit message on each button click (I don't like the first suggestion, I click again for a new one.) 

At the moment, if I don't like the first result, I can't do anything with it it just returns the same every time.

HadesArchitect commented:
My first attempt to use it went not great, autosuggested commit message is not fitting.

1. Suggested
<img width="262" alt="image" src="https://github.com/user-attachments/assets/6073ab03-4a2d-417e-8ced-3aa9c5093802" />
2. Generated in a task: 
<img width="586" alt="image" src="https://github.com/user-attachments/assets/7c10af29-0623-4d8f-8bbe-f47067d5ed63" />

Suggested one is wrong - these files weren't added, but modified. I can't regenerate it, so needed to copy-paste one from the task.

Commit:
♻️ refactor(commit-message): optimize Git service integration

- rename gitUtils to gitService for consistency
- move active repository logic to GitExtensionService
- ensure active repository is fetched before operations
- update gatherStagedChanges to return structured data

✅ test(commit-message): improve test coverage for Git service changes

- update tests to reflect new Git service methods
- mock gatherStagedChanges to return structured changes data
- ensure tests handle no repository and no staged changes scenarios

Commit:
refactor(commit-message): improve git service and testing

Refactor the GitExtensionService to use direct shell commands for git operations
instead of the VS Code Git extension API. This provides more robust and
consistent behavior across different environments.

Updates include:
- `GitExtensionService`: Switched to `execSync` for git commands (`diff`, `status`, `log`).
- `CommitMessageProvider`: Updated to use the new `GitExtensionService` methods for gathering context.
- Test suite: Rewritten to mock `GitExtensionService` and `child_process` directly,
  improving test isolation and reliability.
- Removed unused `vscode.workspace.createFileSystemWatcher` code in `extension.ts`.

Commit:
refactor(git): improve GitExtensionService for better testability

Commit:
💄 style(commit-message): update progress and information messages

- prepend 'Kilo:' to progress and information messages for consistent branding
- enhance user experience by providing clear and branded feedback

Commit:
feat(ai-prompt): enhance AI message regeneration logic

Introduces mechanisms to guide the AI in generating distinct commit messages upon user request for regeneration. This prevents the AI from suggesting identical messages for the same staged changes.

- Adds state to `CommitMessageProvider` to store the last generated message and its associated Git context.
- Implements a method to clear this tracking state, ensuring fresh starts when appropriate.
- Modifies the prompt construction to include explicit instructions for the AI to produce a novel message when a previous one for the same context exists.
- Includes new unit tests to validate the system's ability to prompt for different messages and to confirm the tracking reset functionality.

hassoncs commented:
Great feedback, @HadesArchitect ! I added a couple commits to address your points. 
* Better generated, multi-line messages
* Generate a different message when clicking the button again
* .kilocode/rules can be used to override/alter the prompt passed to the commit message generator

Let me know if this works better for you! I'm still experimenting with it locally.

Commit:
Revert extension changes

Commit:
♻️ refactor(commit-message): localize commit message strings

- replace hardcoded strings with localized message keys
- ensure all commit message related outputs use internationalization

✅ test(commit-message): enhance test coverage for commit message provider

- add mocks for dependencies in commit message provider tests
- verify commit message generation flow and error handling
- ensure proper handling of AI responses with code blocks and formatting

Commit:
Register provider

Commit:
🐛 fix(commit-message): correct git command usage and context handling

- change "git diff --staged" to "git diff --cached" for accuracy
- fix handling of status codes and file paths in GitExtensionService
- update context generation to improve commit message creation

✅ test(commit-message): add tests for no staged changes scenarios

- add test for no staged changes to verify information message display
- add test for empty staged changes array to ensure proper handling

Commit:
feat(i18n): add new messages for commit message provider activation

Adds new internationalization keys for messages related to the activation and registration of the commit message provider. This includes messages for successful registration and activation failures, improving user feedback and debugging.
